pv_site_api/app.py

```diff
@@ -0,0 +1,435 @@
+"""Main API Routes"""
+import os
+from typing import Callable
+
+import pandas as pd
+import sentry_sdk
+import structlog
+from fastapi import APIRouter, Depends, FastAPI, HTTPException
+from fastapi.middleware.cors import CORSMiddleware
+from fastapi.openapi.utils import get_openapi
+from fastapi.responses import FileResponse, JSONResponse
+from pvlib import irradiance, location, pvsystem
+from pvsite_datamodel.read.site import get_all_sites, get_site_by_uuid
+from pvsite_datamodel.read.status import get_latest_status
+from pvsite_datamodel.sqlmodels import ClientSQL, SiteSQL
+from pvsite_datamodel.write.generation import insert_generation_values
+from sqlalchemy.orm import Session
+
+import pv_site_api
+
+from ._db_helpers import (
+    does_site_exist,
+    get_forecasts_by_sites,
+    get_generation_by_sites,
+    site_to_pydantic,
+)
+from .config import Config
+from .fake import (
+    fake_site_uuid,
+    make_fake_forecast,
+    make_fake_pv_generation,
+    make_fake_site,
+    make_fake_status,
+)
+from .pydantic_models import (
+    ClearskyEstimate,
+    Forecast,
+    MultiplePVActual,
+    PVSiteAPIStatus,
+    PVSiteMetadata,
+    PVSites,
+)
+from .redoc_theme import get_redoc_html_with_theme
+from .utils import get_yesterday_midnight
+
+logger = structlog.stdlib.get_logger()
+
+
+def traces_sampler(sampling_context):
+    """
+    Filter tracing for sentry logs.
+
+    Examine provided context data (including parent decision, if any)
+    along with anything in the global namespace to compute the sample rate
+    or sampling decision for this transaction
+    """
+
+    if "error" in sampling_context["transaction_context"]["name"]:
+        # These are important - take a big sample
+        return 1.0
+    elif sampling_context["parent_sampled"] is True:
+        # These aren't something worth tracking - drop all transactions like this
+        return 0.0
+    else:
+        # Default sample rate
+        return 0.05
+
+
+def _add_cors(app: FastAPI, origins: list[str]) -> None:
+    app.add_middleware(
+        CORSMiddleware,
+        allow_origins=origins,
+        allow_credentials=True,
+        allow_methods=["*"],
+        allow_headers=["*"],
+    )
+
+
+def _create_site_router(*, is_fake: bool, get_session: Callable, auth: Callable) -> None:
+    router = APIRouter(prefix="/sites", dependencies=[Depends(auth)])
+
+    @router.get("/", response_model=PVSites)
```

Comment:
The code is the same as before, but it is wrapped into function so that the routes are not defined at import-time, but only when we call the function. This way we can easily inject different dependencies when testing.

Reply:
hmm, I understand why you did this, but its not very neat, and I havent seen any other FastAPIs use this. 

[This](https://github.com/openclimatefix/nowcasting_api/blob/main/src/tests/test_gsp.py#L27) is how I ovverride the session in the other api

Could we do something similar for `auth`? Another way I got round it [was](https://github.com/openclimatefix/nowcasting_api/blob/main/src/auth_utils.py#L24) if `AUTH0_DOMAIN` and `AUTH0_API_AUDIENCE` not set return None - this seemed to get round auth when testing. 

I wonder if you can do something similar?

Reply:
We can use `dependency_override` too, it's just not needed if you inject it like I did.

I can put it back as it was before. This here was how I know to properly inject dependencies, in particular for tests. My main issue is with things like `load_dotenv()` flat in the file (which means that in your tests you'll use your non-test local `.env`, which can be unpredictable) and with all the global objects that then need to be monkeypatched. Injecting them is simpler!

It turns out that fastAPI has a [way to deal with configs][1], and it's to setup the equivalent of my `Config` object as a "Dendency" (using `Depends()`). It feels hackier to me but since it's their recommended approach we can go with that. It fixes the main issues while keeping the code closer to the fastAPI documentation.


[1]: https://fastapi.tiangolo.com/advanced/settings/#settings-in-a-dependency